File: js/46.permutations.js
/*
 * @lc app=leetcode id=46 lang=javascript
 *
 * [46] Permutations
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const res = [];
    const list = [];
    // [], [], []
    helper(res, list, nums);
    function helper(res, list, pool){
        if(pool.length ===  0){
            res.push(list);
        }
        for(const num of pool){
            //there is already a new list coming
            // do not use filter : incase the same case happening here....

            helper(res, [...list, num], pool.filter(x=> num != x););
        }
    }
    // function helper(res, list, nums){
    //     if(list.length === nums){
    //         res.push([...list]);
    //     }
    //     for(let num of nums){
    //         // array using include
    //         // set map has // java; contain/ containsKey
    //         if(!list.includes(num)){
    //             nums
    //         }
    //     }
    // }
    return res;
};
// @lc code=end

File: java/46.permutations.java
import java.util.*;

/*
 * @lc app=leetcode id=46 lang=java
 *
 * [46] Permutations
 */

// @lc code=start
class Solution {

    Map<Integer, Integer> hash;
    List<Integer> path;
    List<List<Integer>> res;

    public List<List<Integer>> permute(int[] nums) {
        hash = new HashMap<Integer, Integer>(nums.length);
        for (int i : nums) {
            hash.put(i, 0);
        }
        path = new ArrayList<Integer>();
        res = new ArrayList<List<Integer>>();

        findPermute(nums, 0);
        return res;
    }

    void findPermute(int[] nums, int start) {
        if (start == nums.length) {
            res.add(new ArrayList<Integer>(path));
        }
        for (int i = 0; i < nums.length; i++) {
            if (hash.get(nums[i]) == 0) {
                path.add(nums[i]);
                hash.put(nums[i], 1);
                findPermute(nums, start + 1);
                hash.put(nums[i], 0);
                path.remove(path.size() - 1);
            }
        }
    }

}
// @lc code=end
